src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+
+            _codeDefinitionWindowService.SetContext(locations);
+        }
+
+        private async Task<ImmutableArray<CodeDefinitionWindowLocation>> GetContextFromPoint(SnapshotPoint? pointInRoslynSnapshot, TaskScheduler foregroundTaskScheduler, CancellationToken cancellationToken)
+        {
+            // TODO: Does this allocate too many tasks - should we switch to a queue like the classifier uses?
+            await Task.Delay(TimeSpan.FromMilliseconds(250), cancellationToken).ConfigureAwait(false);
+            var document = pointInRoslynSnapshot?.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+            }
+
+            var workspace = document.Project.Solution.Workspace;
+            if (!document.SupportsSemanticModel)
+            {
+                var goToDefinitionService = document.GetLanguageService<IGoToDefinitionService>();
+                var navigableItems = await goToDefinitionService.FindDefinitionsAsync(document, pointInRoslynSnapshot.Value.Position, cancellationToken).ConfigureAwait(false);
+                if (navigableItems != null)
+                {
+                    var navigationService = workspace.Services.GetService<IDocumentNavigationService>();
+
+                    var builder = new ArrayBuilder<CodeDefinitionWindowLocation>();
+                    foreach (var item in navigableItems)
+                    {
+                        if (navigationService.CanNavigateToPosition(workspace, item.Document.Id, item.SourceSpan.Start))
+                        {
+                            var text = await item.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                            var linePositionSpan = text.Lines.GetLinePositionSpan(item.SourceSpan);
+                            builder.Add(new CodeDefinitionWindowLocation(item.DisplayString, item.Document.FilePath, linePositionSpan.Start.Line, linePositionSpan.Start.Character));
+                        }
+                    }
+
+                    return builder.ToImmutable();
+                }
+
+                return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+            }
+            else
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = SymbolFinder.FindSymbolAtPosition(
+                    semanticModel,
+                    pointInRoslynSnapshot.Value.Position,
+                    workspace, bindLiteralsToUnderlyingType: true,
+                    cancellationToken: cancellationToken);
+
+                if (symbol == null)
+                {
+                    return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+                }
+
+                symbol = symbol.GetOriginalUnreducedDefinition();
+                var project = document.Project;
```

Comment:
It seems like this is leaving a gotcha where "project" and "document" may be from entirely different workspaces.